src/include/homestore/index/index_internal.hpp

```diff
@@ -63,31 +63,61 @@ enum class index_buf_state_t : uint8_t {
 };
 
 ///////////////////////// Btree Node and Buffer Portion //////////////////////////
-class IndexBuffer;
+struct IndexBuffer;
+struct IndexBufferGroup;
 typedef std::shared_ptr< IndexBuffer > IndexBufferPtr;
+typedef std::shared_ptr< IndexBufferGroup > IndexBufferGroupPtr;
 
 struct IndexBuffer {
     uint8_t* m_node_buf{nullptr};                            // Actual buffer
     index_buf_state_t m_buf_state{index_buf_state_t::CLEAN}; // Is buffer yet to persist?
     BlkId m_blkid;                                           // BlkId where this needs to be persisted
-    IndexBufferPtr m_next_buffer{nullptr};                   // Next buffer in the chain
-    // Number of leader buffers we are waiting for before we write this buffer
-    sisl::atomic_counter< int > m_wait_for_leaders{0};
+    IndexBufferGroupPtr m_buf_group;                         // Buffer group to which it belongs.
 
     IndexBuffer(BlkId blkid, uint32_t buf_size, uint32_t align_size);
+    ~IndexBuffer();
 
     BlkId blkid() const { return m_blkid; }
     uint8_t* raw_buffer() { return m_node_buf; }
 
     bool is_clean() const { return (m_buf_state == index_buf_state_t::CLEAN); }
+    std::string to_string() const {
+        return fmt::format("IndexBuffer {} blkid={} state={} node_buf={}",
+                           reinterpret_cast< void* >(const_cast< IndexBuffer* >(this)), m_blkid.to_integer(),
+                           static_cast< int >(m_buf_state), static_cast< void* >(m_node_buf));
+    }
+};
+
+struct IndexBufferGroup {
+    // First and second index buffers are flushed in order.
+    std::array< IndexBufferPtr, 2 > m_bufs;
+    IndexBufferGroupPtr m_next_group{nullptr}; // Next buffer group in the chain
+    // Number of leader group buffers we are waiting for before we write this buffer
+    sisl::atomic_counter< int > m_wait_for_leaders{0};
+    int32_t m_cur_indx{0};
```

Comment:
So is the idea that we use m_cur_indx for now and come up with different plan to reduce this size?

Reply:
Removed the m_cur_indx and using future then now. There is a todo, currently parent and child and put in separate index buffer group.